### PR TITLE
ATDM: ats2: Correctly build on compute node and run ctest from launch node (CDOFA-102)

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This ATS-2 system is very rigid about where you build and run from.  I fixed
the documentation for how to interactively allocate a compute node and run
builds and tests on this system.  I drivers to run correctly on this system.

For motivation, see [CDOFA-102](https://sems-atlassian-srn.sandia.gov/browse/CDOFA-102).

## How was this tested?

Now to run the full Trilinos test suite for all of the builds:

```
$ ssh vortex

$ /vscratch1/rabartl/Trilinos.base/BUILDS/VORTEX/CTEST_S/

$ time ./ctest-s-local-test-driver.sh all

***
*** ./ctest-s-local-test-driver.sh
***

ATDM_TRILINOS_DIR = '/home/rabartl/Trilinos.base/Trilinos'

Load some env to get python, cmake, etc ...

Hostname 'vortex60' matches known ATDM host 'vortex' and system 'ats2'
Setting compiler and build options for build-name 'default'
Using ats2 compiler stack GNU-7.3.1_SPMPI-2019.06.24 to build DEBUG code with Kokkos node type SERIAL

Running builds:
    ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_opt
    ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_dbg
    ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_opt
    ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_dbg

Fri Mar 13 18:01:39 MDT 2020

Running Jenkins driver Trilinos-atdm-ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_opt.sh ...

    See log file Trilinos-atdm-ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_opt/smart-jenkins-driver.out

real    19m8.685s
user    0m1.090s
sys     0m0.444s

99% tests passed, 1 tests failed out of 2114

Fri Mar 13 18:20:48 MDT 2020

Running Jenkins driver Trilinos-atdm-ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_dbg.sh ...

    See log file Trilinos-atdm-ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_dbg/smart-jenkins-driver.out

real    92m13.791s
user    0m1.323s
sys     0m0.676s

99% tests passed, 1 tests failed out of 2089

Fri Mar 13 19:53:01 MDT 2020

Running Jenkins driver Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_opt.sh ...

    See log file Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_opt/smart-jenkins-driver.out

real    348m15.074s
user    0m1.482s
sys     0m0.952s

100% tests passed, 0 tests failed out of 2232
99% tests passed, 1 tests failed out of 2232

Sat Mar 14 01:41:17 MDT 2020

Running Jenkins driver Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_dbg.sh ...

    See log file Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_dbg/smart-jenkins-driver.out

real    453m43.500s
user    0m1.536s
sys     0m0.738s

99% tests passed, 1 tests failed out of 2179
99% tests passed, 1 tests failed out of 2179

Sat Mar 14 09:15:00 MDT 2020

Done running all of the builds!

real    913m22.332s
user    0m6.096s
sys     0m2.858s

```

That posted to CDash as shown here:

* https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-03-13&end=2020-03-14&filtercount=3&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-ats2&field3=label&compare3=63&value3=Panzer

Amazingly, there were no mass failures with jsrun.  This must be a fluke.

The failing tests in these builds can be seein this this query:

* https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-03-13&end=2020-03-14&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-ats2&field2=buildname&compare2=66&value2=-exp&field3=status&compare3=61&value3=failed&field4=testoutput&compare4=94&value4=OPAL%20ERROR%3A%20Unreachable

which shows the failing tests:

Build Name | Test Name |  Status | Time | Details
-- | -- | -- | -- | -- |
Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_opt-exp | KokkosCore_UnitTest_Cuda_MPI_1 | Failed | 49s 450ms | Completed (Failed)
Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_opt_cuda-aware-mpi-exp | KokkosCore_UnitTest_Cuda_MPI_1 | Failed | 1m 4s 20ms | Completed (Failed)
Trilinos-atdm-ats2-gnu-7.3.1-spmpi-2019.06.24_serial_static_dbg-exp | ROL_example_tempus_example_parabolic_modeleval_forward-adjoint_MPI_1 | Failed | 1m 4s 580ms | Completed (Failed)
Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_dbg-exp | ROL_example_tempus_example_parabolic_modeleval_forward-adjoint_MPI_1 | Failed | 1m 4s 630ms | Completed (Failed)
Trilinos-atdm-ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_dbg_cuda-aware-mpi-exp | ROL_example_tempus_example_parabolic_modeleval_forward-adjoint_MPI_1 | Failed | 1m 4s 970ms | Completed (Failed)

We have already seen these failing tests and have issue trackers:

* https://github.com/trilinos/Trilinos/issues/6799: KokkosCore_UnitTest_Cuda_MPI_1 
* https://github.com/trilinos/Trilinos/issues/5545: ROL_example_tempus_example_parabolic_modeleval_forward-adjoint_MPI_1

So these testing results look excellent.

### CODE REVIEW

**Comment by e10harvey** [comment]
@bartlettroscoe: Please let me know when this is ready for review.

**Comment by bartlettroscoe** [comment]
> Please let me know when this is ready for review.

@e10harvey, it is ready to review.

@jjellio, would you also be willing to look at the updated documentation at:

* https://github.com/trilinos/Trilinos/blob/76ad1e4b0bbdb483d68ce877876f754039db315a/cmake/std/atdm/README.md#ats-2

?

File: cmake/std/atdm/README.md [hunk 1]
```diff
@@ -1050,70 +1050,116 @@ $ bsub -x -Is -n 20 \
...
+```
+$ lalloc 1 -W 4:00 make NP=32
+```
+
+One can also directly run the test suite
+from the login node using:
 
 ```bash
-$ ctest -j4
+$ busub -Is -W 2:00 ctest -j4
```
Comment (e10harvey) [inline cmake/std/atdm/README.md]: typo here, busub should be bsub

File: cmake/std/atdm/README.md [hunk 2]
```diff
@@ -1050,70 +1050,116 @@ $ bsub -x -Is -n 20 \
...
+The MPI test tests in Trilinos are actually run by a wrapper script
+`trilinos_jsrun` which calls the `jsrun` command and modifies the input
+arguments to accommodate the MPI test suite in Trilinos (see the
+implementation of the script `trilinos_jsrun` for details on what it does).
+By default, the script `trilinos_jsrun` will set `export
+TPETRA_ASSUME_CUDA_AWARE_MPI=0` if `TPETRA_ASSUME_CUDA_AWARE_MPI` is unset in
+the environment.  Therefore, by default, the tests are run without CUDA-aware
+MPI on this system.
 
 To explicitly **disable CUDA-aware MPI** when running the test suite, set the
```
Comment (e10harvey) [inline cmake/std/atdm/README.md]: "..., from the launch node set the"

File: cmake/std/atdm/README.md [hunk 3]
```diff
@@ -1050,70 +1050,116 @@ $ bsub -x -Is -n 20 \
...
+environment variable and run as:
 
 ```bash
 $ export TPETRA_ASSUME_CUDA_AWARE_MPI=0
-$ ctest -j4
+$ lrun -n 1 ctest -j4
 ```
 
-and to explicitly **enable CUDA-aware MPI** when running the test suite set:
+To explicitly **enable CUDA-aware MPI** when running the test suite set run
```
Comment (e10harvey) [inline cmake/std/atdm/README.md]: "To explicitly **enable CUDA-aware MPI** when running the test suite, from the launch node run"

File: cmake/std/atdm/ats2/environment.sh [hunk 4]
```diff
@@ -258,11 +255,65 @@ export F77=mpifort
...
+
+
+#
+# Define functions for running on compute nodes
+#
+
+
+function atdm_ats2_get_allocated_compute_node_name() {
+  if [[ "${LSB_HOSTS}" != "" ]] ; then
+    echo ${LSB_HOSTS} | cut -d' ' -f2
```
Comment (e10harvey) [inline cmake/std/atdm/ats2/environment.sh]: Same as above.

File: cmake/std/atdm/common/define_run_on_ats2_compute_node_func.sh [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+  echo
+  echo "Finished running ${script_to_run}!"
+  echo
+
+}
+
+export -f atdm_run_script_on_compute_node
+
+# NOTE: The above function is implemented in this way using 'BSUB' to improve
```
Comment (e10harvey) [inline cmake/std/atdm/common/define_run_on_ats2_compute_node_func.sh]: Consider removing or updating this comment to reflect the changes above.

File: cmake/std/atdm/ctest-s-local-test-driver.sh [hunk 6]
```diff
@@ -301,7 +301,7 @@ for build_name_body in ${ATDM_ARRAY_OF_BUILDS[@]} ; do
     &> smart-jenkins-driver.out
 
   echo
-  grep "failed out of" smart-jenkins-driver.out
+  grep "failed.*out.*of" smart-jenkins-driver.out
```
Comment (e10harvey) [inline cmake/std/atdm/ctest-s-local-test-driver.sh]: Was this `grep` missing failures or did the contents of smart-jenkins-driver.out change?

Comment (bartlettroscoe) [reply cmake/std/atdm/ctest-s-local-test-driver.sh]: > Was this grep missing failures or did the contents of smart-jenkins-driver.out change?

@e10harvey, with the updated CTest 3.16 using color output, it was missing the grep of any test results.  This works now with all versions of CTest, with or without color.

File: cmake/std/atdm/common/define_run_on_ats2_compute_node_func.sh (hunk 5 above)
Comment (bartlettroscoe) [reply cmake/std/atdm/common/define_run_on_ats2_compute_node_func.sh]: > Consider removing or updating this comment to reflect the changes above.

@e10harvey, yes, that I a good suggestion.  Actually, that function and file is not needed anymore (it was an earlier iteration) so I will remove it.

File: cmake/std/atdm/ats2/environment.sh (hunk 4 above)
Comment (bartlettroscoe) [reply cmake/std/atdm/ats2/environment.sh]: `ats2-cuda-10.1.243-gnu-7.3.1-spmpi-2019.06.24_static_opt` returns the empty string if you are not on a compute node.  For example:

```
[rabartl@vortex60 Trilinos (cdofa-102-ats2-fix-run-on-compute-node)]$ . cmake/std/atdm/load-env.sh cuda
Hostname 'vortex60' matches known ATDM host 'vortex' and system 'ats2'
Setting compiler and build options for build-name 'cuda'
Using ats2 compiler stack CUDA-10.1.243_GNU-7.3.1_SPMPI-2019.06.24 to build DEBUG code with Kokkos node type CUDA

[rabartl@vortex60 Trilinos (cdofa-102-ats2-fix-run-on-compute-node)]$ echo "'$(atdm_ats2_get_allocated_compute_node_name)'" 
''

[rabartl@vortex60 Trilinos (cdofa-102-ats2-fix-run-on-compute-node)] $ bsub -Is bash
Job <111821> is submitted to default queue <normal>.
<<Waiting for dispatch ...>>
<<Starting on vortex59>>
Don't load the sems modules on 'vortex'!
[rabartl@vortex59 Trilinos]$ echo "'$(atdm_ats2_get_allocated_compute_node_name)'" 
'vortex12'
```

I will add documentation on this function in my next commit.

**Comment by bartlettroscoe** [comment]
@e10harvey, I pushed updated commits that should hopefully address the outstanding issues.  See commit 237d512.

Can I get a final review so we can merge?

File: cmake/std/atdm/ats2/environment.sh (hunk 4 above)
Comment (e10harvey) [reply cmake/std/atdm/ats2/environment.sh]: Consider `LSB_HOSTS` changing format such that either an an empty or non-empty string is returned on all nodes.

**Comment by e10harvey** [comment]
> @e10harvey, I pushed updated commits that should hopefully address the outstanding issues. See commit [237d512](https://github.com/trilinos/Trilinos/commit/237d512cdf5b3bdda12e4ef403461eb71e2345f7).

Thanks for the quick turn around. I still don't follow: https://github.com/trilinos/Trilinos/pull/7020/files#r393121521. Other than that, I'd like the asserts from (https://github.com/trilinos/Trilinos/pull/7020#pullrequestreview-375340909) added in an effort to save time going forward.

**Comment by bartlettroscoe** [comment]
>  I still don't follow: https://github.com/trilinos/Trilinos/pull/7020/files#r393121521

@e10harvey, let me add a more detailed comment and see if that helps.

> Other than that, I'd like the asserts from (#7020 (review)) added in an effort to save time going forward.

Can you create a commit with what you are suggesting, test it on 'vortex' and then push it to this branch/PR?

**Comment by jjellio** [comment]
> > Please let me know when this is ready for review.
> 
> @e10harvey, it is ready to review.
> 
> @jjellio, would you also be willing to look at the updated documentation at:
> 
> * https://github.com/trilinos/Trilinos/blob/76ad1e4b0bbdb483d68ce877876f754039db315a/cmake/std/atdm/README.md#ats-2
> 
> ?

Ross, When I've been using Trilinos' ENV scripts on compute nodes, I've found that recommending folks do `load-env.sh cuda-debug <machine_name>-....tokens`.  That in general seems to ensure that the scripts always know the proper hostname. (e.g., Mutrino where the node names do not have 'mutrino' in it)

I just worked with someone on getting EMPIRE builds on Mutrino, and the prior was an elegant way to make sure things work as intended

There is typo: (busub) - I can't make live edits into your fork (edits through github.com's IDE)
```
One can also directly run the test suite from the login node using:

$ busub -Is -W 2:00 ctest -j4
```

**Comment by bartlettroscoe** [comment]
@e10harvey, please see updated comments in the driver scripts in commit 49e800e.  Hopefully, if one understands the approach to allocating, building, and testing shown in:

* https://github.com/bartlettroscoe/Trilinos/tree/cdofa-102-ats2-fix-run-on-compute-node/cmake/std/atdm#ats-2

then this approach will make sense?

**Comment by bartlettroscoe** [comment]
> When I've been using Trilinos' ENV scripts on compute nodes, I've found that recommending folks do load-env.sh <machine_name>-....tokens. That in general seems to ensure that the scripts always know the proper hostname. (e.g., Mutrino where the node names do not have 'mutrino' in it)

@jjellio, that may just be a problem with missing logic for the current 'mutrino' ATDM Trilinos env.  I did not realize that you could configure and build directly on the compute nodes on ATS-1 'mutrino'.  We will have to fix this as part of [ATDV-271](https://sems-atlassian-srn.sandia.gov/browse/ATDV-271).

Otherwise, adding 'vortex' to the build name for this 'ats2' env is not going to help at all.

File: cmake/std/atdm/ats2/environment.sh (hunk 4 above)
Comment (bartlettroscoe) [reply cmake/std/atdm/ats2/environment.sh]: > Consider LSB_HOSTS changing format such that either an an empty or non-empty string is returned on all nodes.

@e10harvey, if that happens, we will update the function `atdm_ats2_get_allocated_compute_node_name` do deal with that.  Currently, that function is only used to get the name of the compute node and set it to `CTEST_SITE` for Experimental builds.  So there is little risk if the format of `LSB_HOSTS` changes.  (And, again, we have a function that encapsulates this.)

**Comment by e10harvey** [comment]
> Can you create a commit with what you are suggesting, test it on 'vortex' and then push it to this branch/PR?

If you decide against addressing this here, please open a separate issue.

**Comment by bartlettroscoe** [comment]
> Other than that, I'd like the asserts from (#7020 (review)) added in an effort to save time going forward.

@e10harvey, I just added commit 459d218.  Can you please test these out on 'vortex' trying to do the dumbest things you can think of and see if they are all caught by these checks?  I tested this locally in many different scenarios and it seems to catch all of the invalid usages of these scripts and hopefully gives the user good feedback about what the problem is.

**Comment by bartlettroscoe** [comment]
> Please move the assertions from 459d218 into a common function: assert_node_type(expected_node_type, actual_node_type, expected_hostname) in Trilinos/cmake/std/atdm/common

@e10harvey, I worry about generalizing this until we have a specific case to generalize to (like on another system).   We don't have any other systems like ATS-2 currently where this would even make sense yet.  Also, I worry about being so generic that it makes the error messages less useful.

The Agile approach says to keep it simple until you need to address duplication (i.e. the "fool me once" approach described in ["Agile Software Development" by Bob Martin](https://bartlettroscoe.github.io/reading-list/#agile_software_development_2003)).   Otherwise, we are just guessing (which is bad).

**Comment by bartlettroscoe** [comment]
@e10harvey, I added commit bcf20fb to address review comments.  (I would like to old off on creating a function like `assert_node_type()` until we have a clear reuse case).

**Comment by bartlettroscoe** [comment]
> Code is duplicated at https://github.com/trilinos/Trilinos/pull/7020/files#diff-e0c12389d0389b62da08aac6419d5555R18-R31 and https://github.com/trilinos/Trilinos/pull/7020/files#diff-458dd950017f41243302957e239371fcR12-R25. To minimize code duplication and maintenance going forward, it would be best to just consolidate this code into assert_node_type in our common directory as part of this PR.

@e10harvey, can you create a follow-up PR after this PR merges with your idea for how to do this and what the error messages would like like?

**Pull request merged**